src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,286 @@
+    end,
+  smul_mem' := begin intros c f hf,
+    split,
+    { exact mdifferentiable_smul _ hf.hol },
+    { exact (weakly_modular_submodule k Γ).smul_mem' _ hf.transf },
+    { intro A,
+      rw slash_action.smul_action,
+      apply ((bounded_at_im_infty_subalgebra ℂ).smul_mem (hf.infinity A))}
+    end}
+
+localized "notation `M`:= space_of_mod_forms_of_weight_and_level " in modular_forms
+
+/-- This is the space of cuspforms of weigth `k` and level `Γ` -/
+def space_of_cusp_forms_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2,ℤ)) : submodule ℂ (ℍ → ℂ) :=
+{ carrier := is_cusp_form_of_weight_and_level k Γ,
+  zero_mem' := by apply zero_cusp_form,
+  add_mem' := begin intros a b ha hb,
+    split,
+    { exact mdifferentiable_add ha.hol hb.hol },
+    { exact (weakly_modular_submodule k Γ).add_mem' ha.transf hb.transf },
+    { intro A,
+      rw slash_action.add_action,
+      apply ((zero_at_im_infty_submodule ℂ).add_mem' (ha.infinity A) (hb.infinity A))}
+    end,
+  smul_mem' :=begin intros c f hf,
+    split,
+    { exact mdifferentiable_smul _ hf.hol },
+    { exact (weakly_modular_submodule k Γ).smul_mem' _ hf.transf },
+    { intro A,
+      rw slash_action.smul_action,
+      apply (zero_at_im_infty_submodule ℂ).smul_mem' c (hf.infinity A)},
+    end}
+
+localized "notation `S`:= space_of_cusp_forms_of_weight_and_level" in modular_forms
+
+/--The product of two modular forms is a modular form whose weight is the sum of the weights-/
+lemma mul_modform (k_1 k_2 : ℤ) (Γ : subgroup SL(2,ℤ)) (f g : ℍ → ℂ)
+  (hf : f ∈ M k_1 Γ) (hg : g ∈ M k_2 Γ) : f * g ∈ (M (k_1 + k_2) Γ) :=
+begin
+  refine ⟨mdifferentiable_mul hf.1 hg.1, mul_modular _ _ _ _ _ hf.2 hg.2, λ A, _⟩,
+  rw [slash_mul_SL2 k_1 k_2 A f g],
+  exact prod_of_bounded_is_bounded (hf.infinity A) (hg.infinity A),
+end
+
+/-! Constant functions are modular forms of weight 0. -/
+section const_mod_form
+
+/-- A modular form of weight zero. -/
+def const_one_form : ℍ → ℂ := 1
+
+/-- The constant function is bounded at infinity. -/
+lemma const_one_form_is_bound : is_bounded_at_im_infty const_one_form :=
+@asymptotics.is_O_const_const _ _ ℂ _ _ 1 _ one_ne_zero _
+
+/-- The constant function 1 is invariant under any subgroup of `SL(2,ℤ)`. -/
+lemma const_one_form_is_invar (A : SL(2,ℤ)) : const_one_form ∣[(0 : ℤ), A] = const_one_form :=
+begin
+  rw [slash_action_eq_slash', const_one_form],
+  have : (((↑ₘ(A : GL(2,ℝ)⁺)).det) : ℝ) = 1,
+  { simp only [coe_coe,
+      matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix,
+      matrix.special_linear_group.det_coe], },
+  funext,
+  rw [slash, zero_sub, this],
+  simp only [pi.one_apply, of_real_one, one_zpow, mul_one, neg_zero', zpow_zero],
+end
+
+/-- The constant function 1 is modular of weight 0. -/
+lemma const_mod_form : const_one_form ∈ M 0 Γ :=
+{ hol := by { simp_rw const_one_form, apply mdifferentiable_one },
+  transf := by { intro γ, apply const_one_form_is_invar },
```

Comment:
Use `show_term` to find a term proof.

Reply:
I can't quite get it to work for the `infinity` field...